### ISSUE TITLE
shutdown and restart server error

### ISSUE THREAD

**Issue by joyousrabbit** [alexhultman/uWebSockets#158]
Hello, uWebSoeckets is a great work and I use it in my project. I works very well.

However, everytime I Ctrl+C to shutdown my program, then I re-execute my program, the server cannot get initialized and throw error:

terminate called after throwing an instance of 'decltype(nullptr)'
Aborted

I have to wait at least 30 seconds to re-start my program to keep avoid this error. Becasue I must restart the server often for my test, what a waste of time. Could you please kindly show me what's wrong? Thanks.

./main 
terminate called after throwing an instance of 'decltype(nullptr)'
Aborted

**Comment by alexhultman** [comment]
You need to register a signal handler in which you somehow close the server. Closing the server gracefully makes the system not wait this long until enabling a new process to listen to the same port again.

**Comment by alexhultman** [comment]
Or you could change the system TCP settings (I do this). I have forgotten which setting it is but I think it is something like tcp_reuse or tcp_wait or something like that.

**Comment by joyousrabbit** [comment]
Thank your for rapid response. I added a sigInt functoin to stop the server before exit(1) and everything seems OK now!

**Comment by alexhultman** [comment]
No problem, I just woke up so you had luck with the timing :wink:

**Comment by joyousrabbit** [comment]
Opps, the issus returns back. Same error. Please help me. Maybe
server->close(true) is not enough???

In fact, the server->run() is called from a detached thread;

void VehicleServer::startThread(){
  std::lock_guard<std::mutex> lck(goLock);
  if(gogogo!=-1){ // already started or waiting shutdown
   return;
  }

  gogogo = 1;
  std::thread tListener(&VehicleServer::run,this);
  tListener.detach();
  std::thread tPoll(&VehicleServer::poll,this);
  tPoll.detach();
 }

 void VehicleServer::stopThread(){
  std::lock_guard<std::mutex> lck(goLock);
  server->close(true);
  gogogo = 0;
 }

 void VehicleServer::run(){
  std::cout<<"server start"<<std::endl;
  server->run();
  std::cout<<"server stop"<<std::endl;
  gogogo = -1; // finally terminated
 }

Then in my main.cpp, I catch the Ctrl+C signal and call from main thread
vehicleServer->stopThread, which will trigger server->close(true);

From SSH, I see the server was really stoped:

^CCaught signal 2
socket disconnect 192.168.1.16:58663
server stop
pi@raspberrypi:~/workspace/EncodedRoad $ ^C
pi@raspberrypi:~/workspace/EncodedRoad $ ./main
terminate called after throwing an instance of 'decltype(nullptr)'
Aborted

But for some reason, seems resource was not clearly released and block
restart.
I also make main thread waitting 1 second after stop server, but it didn't
work better.

void my_handler(int s){
 switch(s){
  case 2: //CTRL_C_EVENT
   std::cout<<"Caught signal "<<s<<std::endl;
   encodedRoad->stopThread();
   std::this_thread::sleep_for(std::chrono::seconds(1));
   exit(1);
 }
}

int main(){ ...


2016-06-23 15:23 GMT+02:00 Alex Hultman <notifications@github.com>:

**Comment by SEAPUNK** [comment]
On a side note, could you wrap your code in three backticks? It will format the code, and make it much easier to read.

Optionally, you can also add the prefix immediately after the three backticks, like so:

    ```cpp
    // code goes here
    ```

**Comment by alexhultman** [comment]
I can't see the whole code. You need to set master = false on any Server not being run on the main thread. That will make Server::close and Server::upgrade thread safe so that you can load balance connections. The internal instances are not thread safe though, so you cannot send data on a separate thread than the one the server is run on. Not currently at least, not without manually tinkering with the uv_async support.
